src/urllib3/connection.py

```diff
@@ -422,17 +521,20 @@ def connect(self):
             cert = self.sock.getpeercert()
             _match_hostname(cert, self.assert_hostname or server_hostname)
```

Comment:
`self.assert_hostname` is of type `Optional[Union[str,bool]]`. Should users be allowed to pass `True` to `assert_hostname`? If so, `_match_hostname` should then be called with `server_hostname`? The other option I see is for `assert_hostname` to be of type `Optional[Union[Literal['False'], str]]`.